### PR TITLE
Add missing UEFI config options to bfcfg dump

### PR DESCRIPTION
**Pull request by cbabroski**
Add missing UEFI config options to bfcfg dump so the bfcfg dump output can be used to compare defaults across releases. A dump level option has also been added to bfcfg to increase the amount of information dumped like the boot order and public keys/certificates used for secure boot.

RM #3969767

### CODE REVIEW

**Review by jamesphurl — CHANGES_REQUESTED** [review]
Same issue as with edk2.

File: bfcfg [hunk 1]
```diff
@@ -37,19 +37,27 @@ bfcfg_rc=0
...
 efi_global_var_guid=8be4df61-93ca-11d2-aa0d-00e098032b8c
 efi_bfcfg_var_guid=487ff588-fb71-4b19-9100-ebe067aa1af0
 efi_vlan_var_guid=9e23d768-d2f3-4366-9fc3-3a7aba864374
 rshim_efi_mac_sysfs=${efivars}/RshimMacAddr-${efi_global_var_guid}
 pxe_dhcp_class_id_sysfs=${efivars}/DhcpClassId-${efi_global_var_guid}
 bf_bundle_version_sysfs=${efivars}/BfBundleVer-${efi_bfcfg_var_guid}
+sb_custom_mode_sysfs=${efivars}/BfCfgSbCustomMode-${efi_bfcfg_var_guid}
+pass_settings_sysfs=${efivars}/BfCfgPassSettings-${efi_bfcfg_var_guid}
+passwd_sysfs=${efivars}/BfCfgPassword-${efi_bfcfg_var_guid}
+bf_modes_sysfs=${efivars}/BfCfgBfModes-${efi_bfcfg_var_guid}
```
[the remark this answers is not in the document]
Comment (cbabroski) [reply bfcfg:60]: Hey Jim, I ended up removing the legacy password dump. I couldn't reproduce a scenario where the legacy password wasn't all 0s - the PwdCredential variable was either 0 or did not exist (which is similar to the secure password variable when no password is set). If it's correct that the legacy password is only used on older builds (<= 4.5) and gets converted to a secure password on newer builds then I think it's okay to just dump the secure password since the related edk2 changes won't be available on older builds.

Let me know what you think, happy to investigate more and if you have a suggested method to repro a non-zero legacy password in 4.8 I'll try that out. Thanks!

[the remark this answers is not in the document]
Comment (jamesphurl) [reply bfcfg:60]: Thats fine Chris. See no value including the Legacy Password. It only applies downgrading to an older (<= 4.5.x) release., otherwise as you mentioned it won't exist or be all 0's. Only reason I mentioned at all was because the wrong UEFI variable was being used to print it out.

[the remark this answers is not in the document]
Comment (cbabroski) [reply bfcfg:60]: Sounds good, and yeah thanks for bringing that up!

**Comment by kblaiech** [comment]
Are the three commits expected or were you trying to update this pull request?

If you are trying to update this pull request, you can run "git push origin master -f" against your fork. That should update the files in this pull request.

File: bfcfg [hunk 2]
```diff
@@ -687,9 +714,16 @@ sys_cfg()
   sys_cfg_one_byte ${tmp_file} "ENABLE_SPCR" 32 "${SYS_ENABLE_SPCR}"
   sys_cfg_one_byte ${tmp_file} "DISABLE_PCIE" 33 "${SYS_DISABLE_PCIE}"
   sys_cfg_one_byte ${tmp_file} "ENABLE_OPTEE" 34 "${SYS_ENABLE_OPTEE}"
+  sys_cfg_one_byte ${tmp_file} "DISABLE_TMFF" 35 "${SYS_DISABLE_TMFF}"
   sys_cfg_one_byte ${tmp_file} "ENABLE_I2C0" 36 "${SYS_ENABLE_I2C0}"
   sys_cfg_one_byte ${tmp_file} "DISABLE_FORCE_PXE_RETRY" 37 "${SYS_DISABLE_FORCE_PXE_RETRY}"
   sys_cfg_one_byte ${tmp_file} "ENABLE_BMC_FIELD_MODE" 38 "${SYS_ENABLE_BMC_FIELD_MODE}"
+  sys_cfg_one_byte ${tmp_file} "DEFAULT_LARGE_ICMC_LOCK" 43 "${SYS_DEFAULT_LARGE_ICMC_LOCK}"
+  sys_cfg_one_byte ${tmp_file} "DISPLAY_FLAGS" 44 "${SYS_DISPLAY_FLAGS}"
```
Comment (kblaiech) [inline bfcfg:722]: This isn't a real platform configuration; this is toggled through key chain. I don't think we should expose this option.

File: bfcfg [hunk 3]
```diff
@@ -648,22 +655,42 @@ sys_cfg_one_byte()
   fi
 }
 
+sys_cfg_multi_byte()
```
Comment (kblaiech) [inline bfcfg:658]: Could you please leave a comment to indicate that this function only displays the value? It doesn't set it (unlike sys_cfg_one_byte)

I would also keep the number of arguments the same as sys_cfg_one_byte, although the value won't be ignored. Who knows maybe we would need to configure other multi bytes sys config in the future

File: bfcfg [hunk 4]
```diff
@@ -648,22 +655,42 @@ sys_cfg_one_byte()
...
+{
+  local tmp_file=$1
+  local name=$2
+  local offset=$3
+  local num_bytes=$4
+
+  if [ ${dump_mode} -eq 1 ]; then
+    value=0
+    if [ -e ${tmp_file} ]; then
+      value=$(hexdump -s "${offset}" -n "${num_bytes}" -e '"%d" "\n"' "${tmp_file}")
```
Comment (kblaiech) [inline bfcfg:668]: Does it take into account the byte order? Have you tested that?

**Comment by cbabroski** [comment]
> Are the three commits expected or were you trying to update this pull request?
> 
> If you are trying to update this pull request, you can run "git push origin master -f" against your fork. That should update the files in this pull request.

@kblaiech Since Github has no way to track incremental changes on a code review using an ID like gitreview/git-master I'm using multiple commits to make reviewing easier. I'll squash them together before merging so it will look no different on master.

File: bfcfg (hunk 3 above)
Comment (cbabroski) [reply bfcfg:658]: Hey Khalil, I updated the function name and added a comment to make it more clear that this is for dumping.

I left out support for setting multi-byte values because it isn't needed yet and we have to be more careful about correctly handling the input size of the arguments passed to bfcfg when setting variables to avoid overwriting memory in the `SYS_CONFIG` struct. For that reason I also removed the "value" input argument.

In the future it probably makes sense to remove the distinction of `sys_cfg_one_byte` and `sys_cfg_multi_byte` and combine into one common function that can set and dump any number of bytes at a given offset. Didn't want to do that here because it would increase the regression test load for a case that isn't needed yet, but happy to do in the future if we want to add that support.

File: bfcfg (hunk 4 above)
Comment (cbabroski) [reply bfcfg:668]: Hexdump appears to use the machine's native endianness and formats the decimal output correctly. To test I reset my EFI vars, installed the latest Ubuntu release, and verified the values match the expected defaults:
```
sys: LARGE_ICMC_SIZE=512
sys: CE_THRESHOLD=5000
```
Then I modified the values in UEFI menu and verified they appear as expected:
```
 Large ICMC size            [280]
 CE Threshold               [1300]
...
sys: LARGE_ICMC_SIZE=640
sys: CE_THRESHOLD=4864
```
I can update it so the values are displayed in hex though which is what the UEFI menu input requires.

Comment (kblaiech) [reply bfcfg]: Thanks!

File: bfcfg [hunk 5]
```diff
@@ -687,10 +720,15 @@ sys_cfg()
...
   sys_cfg_one_byte ${tmp_file} "ENABLE_OPTEE" 34 "${SYS_ENABLE_OPTEE}"
+  sys_cfg_one_byte ${tmp_file} "DISABLE_TMFF" 35 "${SYS_DISABLE_TMFF}"
   sys_cfg_one_byte ${tmp_file} "ENABLE_I2C0" 36 "${SYS_ENABLE_I2C0}"
   sys_cfg_one_byte ${tmp_file} "DISABLE_FORCE_PXE_RETRY" 37 "${SYS_DISABLE_FORCE_PXE_RETRY}"
   sys_cfg_one_byte ${tmp_file} "ENABLE_BMC_FIELD_MODE" 38 "${SYS_ENABLE_BMC_FIELD_MODE}"
 
+  # Multi-byte variable dump (read only)
+  sys_cfg_dump_multi_byte ${tmp_file} "LARGE_ICMC_SIZE" 39 4
+  sys_cfg_dump_multi_byte ${tmp_file} "CE_THRESHOLD" 45 4
+
```
Comment (kblaiech) [inline bfcfg:731]: I recommend having these calls guarded by a checkpoint on ${dump_mode}, just to improve the readability and avoid confusion with calls used to set the values. Actually it was for that reason I asked to add an extra (unused) parameter and mirror sys_cfg_one_byte but without having to deal with the "set", the 'sys_cfg_dump_multi_byte' routine would handle only the 'dump' and we'll have sort of TBD/TODO tag to deal with enhancements later on.

The bfcfg is not a perfect/clean script and had many contributors/developers adding bunch of commands. One cleaner way is to handle the 'dump' separately, but as you just pointed out we don't want to cause any regression so I suggest either solutions:

Solution #1

```
if [ ${dump_mode} -eq 1 ]; then
  # Multi-byte variable dump (read only)
  sys_cfg_dump_multi_byte ${tmp_file} "LARGE_ICMC_SIZE" 39 4
  sys_cfg_dump_multi_byte ${tmp_file} "CE_THRESHOLD" 45 4
fi
```

Solution #2

```
sys_cfg_multi_byte ${tmp_file} "LARGE_ICMC_SIZE" 39 4 "SYS_LARGE_ICMC_SIZE"
sys_cfg_multi_byte ${tmp_file} "CE_THRESHOLD" 45 4 "SYS_CE_THRESHOLD"
```

What do you think?
(This is a personal preference as I would like to make the script readable and easy for other developers to follow)

File: bfcfg [hunk 6]
```diff
@@ -37,19 +37,26 @@ bfcfg_rc=0
 cfg_file=/etc/bf.cfg
 log_file=/tmp/bfcfg.log
 dump_mode=0
+dump_all_mode=0
```
Comment (lsun100) [inline bfcfg:40]: A more extensive way is to reuse the dump_mode, so later we don't have to add more variables, such as

dumo_mode:
1: existing dump
2: all
3: selective dump to be defined later.

File: bfcfg [hunk 7]
```diff
@@ -648,29 +655,157 @@ sys_cfg_one_byte()
...
+  local num_bytes=$4
+  local value=$5
+
+  if [ ${dump_mode} -eq 0 ]; then
+    return
+  fi
+
+  value=0
+  if [ -e ${tmp_file} ]; then
+    value=$(hexdump -s "${offset}" -n "${num_bytes}" -e '"0x%X" "\n"' "${tmp_file}")
```
Comment (lsun100) [inline bfcfg:675]: Is there any case that the output could be over one line thus the output might be not expected?

File: bfcfg [hunk 8]
```diff
@@ -648,29 +655,157 @@ sys_cfg_one_byte()
...
+# This function reads the BfModes EFI struct from a UEFI variable directly.
+# Below are the offsets defined in UEFI which is 4(fixed header) plus the offset
+# of the variable within the BfModes struct. These offsets are not supposed to
+# change in order to be backward compatible with previous releases.
+#   VARIABLE(Name)       OFFSET(Byte)  SIZE(Byte)
+#   INTERNAL_CPU_MODEL        4            1
+#   HOST_PRIV_LEVEL           5            1
+#   NIC_MODE                  6            1
+#
+sys_cfg_dump_bf_modes()
```
Comment (lsun100) [inline bfcfg:690]: The sys_cfg_xxx() is intended for UEFI sys config operations, while this one is not.
Can we use a different name to avoid confusion, such as sysfs_dump_bf_modes()?

File: bfcfg [hunk 9]
```diff
@@ -648,29 +655,157 @@ sys_cfg_one_byte()
...
+{
+  local tmp_file=/tmp/.bfcfg-sysfs-data
```
Comment (lsun100) [inline bfcfg:692]: 1. same comment here to use a different name.
2. Would it be possible to use a temp directory (check example in bfver line 31-33), and replace all /tmp/xxx with ${TMP_DIR}/xxx?  So it could avoid race condition when multiple people run it is parallel.

File: bfcfg [hunk 10]
```diff
@@ -648,29 +655,157 @@ sys_cfg_one_byte()
...
+    log_msg "sys: failed to find the ${bf_modes_sysfs} EFI variable"
+    return
+  fi
+
+  # shellcheck disable=SC2216
+  yes | cp -f ${bf_modes_sysfs} ${tmp_file}
+  has_change=0
+
+  # shellcheck disable=SC2216
+  yes | cp -f ${bf_modes_sysfs} ${tmp_file}
```
Comment (lsun100) [inline bfcfg:711]: Is this line the same as line 707?

File: bfcfg [hunk 11]
```diff
@@ -648,29 +655,157 @@ sys_cfg_one_byte()
...
+    nic_mode_str="DPU_MODE"
+  elif [ "${nic_mode}" -eq 1 ]; then
+    nic_mode_str="NIC_MODE"
+  else
+    nic_mode_str="UNKNOWN"
+  fi
+
+  echo "sys: INTERNAL_CPU_MODEL=${internal_cpu_model_str}"
+  echo "sys: HOST_PRIV_LEVEL=${host_priv_level_str}"
+  echo "sys: NIC_MODE=${nic_mode_str}"
```
Comment (lsun100) [inline bfcfg:748]: same here, use "sysfs: " instead if "sys: " to differentiate the SysConfig in UEFI?

File: bfcfg [hunk 12]
```diff
@@ -648,29 +655,157 @@ sys_cfg_one_byte()
...
+
+  rm -f ${tmp_file}
```
Comment (lsun100) [inline bfcfg:750]: No need for this line if we use TMP_DIR

File: bfcfg [hunk 13]
```diff
@@ -648,29 +655,157 @@ sys_cfg_one_byte()
...
+}
+
+sys_cfg_dump_pass_settings()
```
Comment (lsun100) [inline bfcfg:753]: Same for this function name

File: bfcfg [hunk 14]
```diff
@@ -648,29 +655,157 @@ sys_cfg_one_byte()
...
+    log_msg "sys: failed to find the ${pass_settings_sysfs} EFI variable"
+    return
+  fi
+
+  # shellcheck disable=SC2216
+  yes | cp -f ${pass_settings_sysfs} ${tmp_file}
+  default_pass_policy=$(hexdump -s 4 -n 1 -e '/1 "%d" "\n"' ${tmp_file})
+  echo "sys: UEFI_DEFAULT_PASS_POLICY_ENABLE=${default_pass_policy}"
+
+  rm -f ${tmp_file}
```
Comment (lsun100) [inline bfcfg:772]: No need if using TMP_DIR

File: bfcfg [hunk 15]
```diff
@@ -648,29 +655,157 @@ sys_cfg_one_byte()
...
 {
   local tmp_file=/tmp/.bfcfg-sysfs-data
 
   has_sys_cfg=0
 
+  if [ ${dump_mode} -eq 1 ]; then
+    sys_cfg_dump_bf_modes
+    sys_cfg_dump_pass_settings
+  fi
+
```
Comment (lsun100) [inline bfcfg:808]: This block seems not related to UEFI SysConfig. Move it to 1425-1430 block below?

File: bfcfg [hunk 16]
```diff
@@ -1043,7 +1245,10 @@ boot_cfg()
   local l4proto l4proto_len
   local oob_mac_addr num oob_vlan_path
 
-  [ $dump_mode -eq 1 ] && return
+  if [ $dump_mode -eq 1 ]; then
```
Comment (lsun100) [inline bfcfg:1248]: A few places use '$dump_mode -eq 1'.  You might want to check if we should use '$dump_mode -ne 0' to cover the '-all' case.

File: bfcfg [hunk 17]
```diff
@@ -1393,14 +1678,15 @@ print_osarg()
...
+  echo "syntax: bfcfg [--help|-h] [--dump|-d] [--all|-a] [--dump-osarg|o] [--part-info|-p] [--cfg2bin|-c <cfg_file>]"
 }
 
 # Parse the arguments.
-options=$(getopt -n bfcfg -o dohpc: -l dump,dump-osarg,help,part-info,cfg2bin: -- "$@")
+options=$(getopt -n bfcfg -o adohpc: -l all,dump,dump-osarg,help,part-info,cfg2bin: -- "$@")
 eval set -- "$options"
 while [ "$1" != -- ]; do
   case $1 in
+    --all|-a) dump_all_mode=1 ;;
```
Comment (lsun100) [inline bfcfg:1689]: Maybe we should add a dependency check here dump_mode != 0, and set dump_mode=2?

File: bfcfg (hunk 6 above)
Comment (cbabroski) [reply bfcfg:40]: I think we still need a new script parameter. Optional param arguments are not POSIX standard (for example being able to pass both `bfcfg -d` with no arguments and also `bfcfg -d 2`.

So two solutions could be:
1. `bfcfg -d <dump_mode>` : required `-d` argument
2. `bfcfg -d -l <dump_level>` : new param (replace `-a|--all` with `-l|--dump-level`)

Since the first option can break backward compatibility (e.g. someone uses `bfcfg -d` with no arguments in a script), I'm leaning towards the second option. Thoughts?

Comment (lsun100) [reply bfcfg:40]: Thanks!, yes #2 seems a lot better

File: bfcfg (hunk 7 above)
Comment (cbabroski) [reply bfcfg:675]: Not sure I fully understand - do you mean if we tried to dump too many bytes that it would look bad (e.g. `sys: VAR=0x1234678901234567809012325366708...`?

`sys_cfg_multi_byte` is meant to be used as private function called by other functions like `sys_cfg` so it doesn't need to handle a variable input from running `bfcfg`, for example.

But we could enforce a max `num_bytes` size in the function or split the output into multiple lines if `num_bytes` was larger than a set threshold.

Not sure it's necessary now though since this is only being used to dump 2-4 bytes (meant mostly for UINT16 and UINT32 sys config vars). What do you think?

File: bfcfg (hunk 8 above)
Comment (cbabroski) [reply bfcfg:690]: Yeah I'm happy to rename, but does this change now that we have `SYS_IN_CPU_MODEL`, `MFG_HOST_PRIV_MODE`, and `MFG_PLAT_MODE` as part of the `ARM_CFG_SUBTYPE_SYS` platform config: https://confluence.nvidia.com/display/SW/BlueField+Platform+Configuration+Blob+Format

@kblaiech thoughts?

File: bfcfg (hunk 10 above)
Comment (cbabroski) [reply bfcfg:711]: Ah yeah, thanks

File: bfcfg (hunk 11 above)
Comment (cbabroski) [reply bfcfg:748]: Same as above - happy to change, but not sure if we want to keep "sys:" given the recent changes to the platform config format (to show they are in the same ARM config)?

File: bfcfg (hunk 12 above)
Comment (cbabroski) [reply bfcfg:750]: The only thing is that `rm -f ${tmp_file}` is already used in many places throughout the bfcfg code so I feel like it may be more confusing when reading through if that line doesn't exist only for a few functions. I could remove all of them, but I'm a little worried about causing regressions...

Does the following guarantee the file will be removed if called inside a function/subshell?
```
trap "rm -rf $TMP_DIR" EXIT
trap "rm -rf $TMP_DIR; exit 1" TERM INT
```
I tried to test it out and saw that the directory was still there after the function exited.

File: bfcfg (hunk 14 above)
Comment (cbabroski) [reply bfcfg:772]: Same question/concern as above

File: bfcfg (hunk 15 above)
Comment (cbabroski) [reply bfcfg:808]: I had moved these two functions into sys_cfg() based on feedback from @kblaiech - I think the code has moved around so lines 1425-1430 may not be pointing to the same place anymore. Were you suggesting to call at the end of bfcfg or somewhere else?

File: bfcfg (hunk 12 above)
Comment (lsun100) [reply bfcfg:755]: The 'trap' command should guarantee the file to be removed I think. So only need to make sure the old tmp_file generated by one function not to be unused by another function incorrectly.

File: bfcfg [hunk 18]
```diff
@@ -2543,6 +2828,16 @@ while [ "$1" != -- ]; do
 done
 shift
 
+# Ensure dump_level is set correctly based on input dump_mode and dump_level.
+if [ ${dump_mode} -eq 0 ]; then
+  dump_level=${DUMP_LEVEL_NONE}
+elif [ ${has_dump_level} -eq 0 ]; then
```
Comment (lsun100) [inline bfcfg:2834]: Looks like the 'has_dump_level' is not necessary. It can be simplified as elif [ -z "${dump_level} ]

File: bfcfg (hunk 8 above)
Comment (cbabroski) [reply bfcfg:699]: Changed to sysfs_dump_bf_modes and sysfs_dump_pass_settings

File: bfcfg (hunk 12 above)
Comment (cbabroski) [reply bfcfg:755]: Removed 'rm -f' for the new bf modes and pass settings dump functions and made sure they use a unique temp file name

File: bfcfg (hunk 11 above)
Comment (cbabroski) [reply bfcfg:753]: Changed to sysfs

Comment (cbabroski) [reply bfcfg:753]: Actually could also use 'misc:' and put into misc_cfg instead?

Comment (lsun100) [reply bfcfg:753]: > Actually could also use 'misc:' and put into misc_cfg instead?

Sure, it also seems good.

**Review by lsun100 — APPROVED** [review]

File: bfcfg (hunk 18 above)
Comment (cbabroski) [reply bfcfg]: done

File: bfcfg (hunk 9 above)
Comment (cbabroski) [reply bfcfg]: The only thing I had to revert just now was the log file located at /tmp/bfcfg.log since that file has to persist after the script exits. All other instances of /tmp/X are using TMP_DIR now though

**Review by lsun100 — APPROVED** [review]
